### PR TITLE
Wrote basic client windowing

### LINKED ISSUE #6: Setup handlers so client can add to its local move list

### PR DESCRIPTION
**Pull request by JustinHuPrime**
Client now has basic windowing set up.

Closes #6

### CODE REVIEW

**Review by JohannCooper — APPROVED** [review]

File: src/client/Main.hs [hunk 1]
```diff
@@ -15,37 +15,48 @@ PARTICULAR PURPOSE. See the GNU General Public License for more details.
...
+  keyboardMouseCallback $= Just (keyboardMouse board boardLock moveList moveListLock)
+  idleCallback $= Just (idle board boardLock)
+  closeCallback $= Just (close s)
+
+  mainLoop
 
 client :: String -> IO ()
 client serverAddr = do
   board <- newIORef [] :: IO (IORef Board)
+  boardLock <- Lock.new
```
Comment (wyndro) [inline src/client/Main.hs:55]: Why do we need locks in the client? Is there more than one thread running in the client or is this for interrupts of some kind?

File: src/client/Main.hs [hunk 2]
```diff
@@ -15,37 +15,48 @@ PARTICULAR PURPOSE. See the GNU General Public License for more details.
...
   moveList <- newIORef [] :: IO (IORef [Move])
-  lock <- Lock.new
+  moveListLock <- Lock.new
```
Comment (wyndro) [inline src/client/Main.hs:57]: Especially this lock. Isn't there only one player modifying the move list in the client? Can you explain what you are thinking we need these for?

File: src/client/Main.hs [hunk 3]
```diff
@@ -15,37 +15,48 @@ PARTICULAR PURPOSE. See the GNU General Public License for more details.
...
+  initBoard <- readBoard s
+  writeIORef board initBoard
+
+  -- TODO: are there any additional options we need to pass?
+  initialDisplayMode $= [DoubleBuffered]
+  -- TODO: probably need to setup openGL versions and such
+  _ <- createWindow "N-Planetary"
+  fullScreen
+
+  displayCallback $= display board boardLock
```
Comment (wyndro) [inline src/client/Main.hs:44]: `$=` is this equivalent to `set!` in ASL?

File: src/client/Main.hs [hunk 4]
```diff
@@ -15,37 +15,48 @@ PARTICULAR PURPOSE. See the GNU General Public License for more details.
...
+  -- TODO: add more input callbacks as needed
+  keyboardMouseCallback $= Just (keyboardMouse board boardLock moveList moveListLock)
+  idleCallback $= Just (idle board boardLock)
+  closeCallback $= Just (close s)
```
Comment (wyndro) [inline src/client/Main.hs:48]: Do these things need to be handed to something? I am missing where these variables are stored. What's their scope?

File: src/client/Main.hs (hunk 1 above)
Comment (JustinHuPrime) [reply src/client/Main.hs:55]: I think there's event handlers - the display callback might (I'm not sure how it works internally) get called while a click is being processed. If they're in different threads, this is necessary.

File: src/client/Main.hs (hunk 2 above)
Comment (JustinHuPrime) [reply src/client/Main.hs:57]: Again, multiple click callbacks might happen at the same time.

File: src/client/Main.hs (hunk 3 above)
Comment (JustinHuPrime) [reply src/client/Main.hs:44]: It's writeIORef as a user-defined operator.

File: src/client/Main.hs (hunk 4 above)
Comment (JustinHuPrime) [reply src/client/Main.hs:48]: They're global IORefs that the GLU library sets up. They already belong to the library.

**Review by wyndro — APPROVED** [review]

**Pull request merged**